### ISSUE TITLE
Issues with Oscillator.type  "custom"

### ISSUE THREAD

**Issue by rtoy** [WebAudio/web-audio-api#311]
In http://webaudio.github.io/web-audio-api/#attributes-19, the type is not allowed to be set to "custom" directly.

There are a few issues here:
 * What should happen if you do set it to "custom"?
 * Since "custom" is a valid value for OscillatorType, this seems to violate the WebIDL spec about setting the type.
 * The text says that setPeriodicWave "can be used". Shouldn't this say "MUST be used"?

**Comment by joeberkovitz** [comment]
I think this may already be dictated by the outcome of #383 which suggests that nothing happens if an invalid value is set for an enum.

**Comment by rtoy** [comment]
Sorry, I don't quite follow.  "custom" is a valid enum for oscillator types.

Chrome currently just ignores you if you try to set the type to custom.  The spec already says you cannot set it directly to custom. I think we just need to clarify that setting it does nothing.

**Comment by padenot** [comment]
I agree with @rtoy, it's a valid value. Gecko currently throws with `InvalidStateError`, and as usual, I prefer explicit (throwing) to implicit.

**Comment by rtoy** [comment]
@padenot I don't quite follow.  Should we throw an error or not when setting the type to custom? I would also prefer throwing.

**Comment by padenot** [comment]
Yeah I was a bit unclear, let me rephrase:

Since it's a valid value, we are not, per spec, mandated to silently ignore it.

I'm proposing we spec that setting it to `"custom"` throws an `InvalidStateError`.

**Comment by rtoy** [comment]
Sounds good.

On Tue, May 12, 2015 at 9:18 AM, Paul Adenot <notifications@github.com>
wrote:

> Yeah I was a bit unclear, let me rephrase:
>
> Since it's a valid value, we are not, per spec, mandated to silently
> ignore it.
>
> I'm proposing we spec that setting it to "custom" throws an
> InvalidStateError.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAudio/web-audio-api/issues/311#issuecomment-101336887>
> .
>



-- 
Ray
